core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -0,0 +1,795 @@
+				     val & QM_FVT_CFG_RDY_BIT, POLL_PERIOD,
+				     POLL_TIMEOUT);
+	if (ret) {
+		EMSG("QM VFT is not ready");
+		return HISI_QM_DRVCRYPT_EBUSY;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status qm_set_xqc_vft(struct hisi_qm *qm,
+					   uint32_t function,
+					   uint32_t base, uint32_t num)
+{
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+	int32_t i = 0;
+
+	if (!num) {
+		EMSG("Invalid sq num");
+		return HISI_QM_DRVCRYPT_EINVAL;
+	}
+
+	for (i = QM_SQC_VFT; i <= QM_CQC_VFT; i++) {
+		ret = qm_set_vft_common(qm, i, function, base, num);
+		if (ret) {
+			EMSG("QM set type%d fail", i);
+			return ret;
+		}
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status qm_get_vft(struct hisi_qm *qm, uint32_t *base,
+				       uint32_t *num)
+{
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+	uint64_t sqc_vft = 0;
+
+	ret = qm_mb(qm, QM_MB_CMD_SQC_VFT, 0, 0, QM_MB_OP_RD);
+	if (ret)
+		return ret;
+
+	sqc_vft = io_read64(qm->io_base + QM_MAILBOX_DATA_ADDR_L);
+	*base = (sqc_vft >> QM_SQC_VFT_START_SQN_SHIFT) & QM_SQC_VFT_BASE_MASK;
+	*num = ((sqc_vft >> QM_SQC_VFT_SQ_NUM_SHIFT) & QM_SQC_VFT_NUM_MASK) + 1;
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static void qp_memory_uninit(struct hisi_qm *qm, uint32_t id)
+{
+	struct hisi_qp *qp = &qm->qp_array[id];
+
+	free(qp->sqe);
+	free(qp->cqe);
+}
+
+static enum hisi_drv_status qp_memory_init(struct hisi_qm *qm, uint32_t id)
+{
+	size_t sq_size = qm->sqe_size * HISI_QM_Q_DEPTH;
+	size_t cq_size = sizeof(struct qm_cqe) * HISI_QM_Q_DEPTH;
+	struct hisi_qp *qp = &qm->qp_array[id];
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+
+	qp->sqe = memalign(HISI_QM_ALIGN128, sq_size);
+	if (!qp->sqe) {
+		EMSG("Fail to malloc sq[%"PRIu32"]", id);
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+	qp->sqe_dma = virt_to_phys(qp->sqe);
+	qp->cqe = memalign(HISI_QM_ALIGN32, cq_size);
+	if (!qp->cqe) {
+		EMSG("Fail to malloc cq[%"PRIu32"]", id);
+		ret = HISI_QM_DRVCRYPT_ENOMEM;
+		goto free_sqe;
+	}
+	qp->cqe_dma = virt_to_phys(qp->cqe);
+
+	qp->qp_id = id;
+	qp->qm = qm;
+	return HISI_QM_DRVCRYPT_NO_ERR;
+
+free_sqe:
+	free(qp->sqe);
+	return ret;
+}
+
+static void qm_memory_uninit(struct hisi_qm *qm)
+{
+	uint32_t i = 0;
+
+	for (i = 0; i < qm->qp_num; i++)
+		qp_memory_uninit(qm, i);
+
+	free(qm->qp_array);
+	free(qm->sqc);
+	free(qm->cqc);
+}
+
+static enum hisi_drv_status qm_memory_init(struct hisi_qm *qm)
+{
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+	size_t sqc_size = 0;
+	size_t cqc_size = 0;
+	size_t qp_size = 0;
+	uint32_t i = 0;
+	int32_t j = 0;
+
+	sqc_size = sizeof(struct qm_sqc) * qm->qp_num;
+	cqc_size = sizeof(struct qm_cqc) * qm->qp_num;
+	qp_size = sizeof(struct hisi_qp) * qm->qp_num;
+
+	qm->sqc = memalign(HISI_QM_ALIGN32, sqc_size);
+	if (!qm->sqc) {
+		EMSG("Fail to malloc sqc");
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+	qm->sqc_dma = virt_to_phys(qm->sqc);
+
+	qm->cqc = memalign(HISI_QM_ALIGN32, cqc_size);
```

Comment:
should both allocated `qm->sqc` and `qm->cqc` be zeroed?

Reply:
Yeah，sqc,cqc,sqe,cqe needs to be used repeatedly, so it is initialized to 0 before it is used.
```c
memzero_explicit(sqc, sizeof(struct qm_sqc));
memzero_explicit(cqc, sizeof(struct qm_cqc));
memzero_explicit(qp->cqe, sizeof(struct qm_cqe) * HISI_QM_Q_DEPTH);
memzero_explicit(sqe, qm->sqe_size);
```

Reply:
Ok. Note you don't need `memzero_explicit()`. `memset(sqe, 0, qm->sqe_size);` is enough.
`memzero_explicit()` is to be used when one needs to prevent compiler from discarding memory writes. See [_string_ext.h:37_](https://github.com/OP-TEE/optee_os/blob/master/lib/libutils/ext/include/string_ext.h#L37-L43).

That said, I don't think you really need to zero the memory before its used, It should be enough to wipe/zero initial content after allocation. Yet, both implementations should do the job.

Reply:
You are right ,using memset is enough.

I just checked,sqc and cqc wipe/zero initial content after allocation is enought, but sqe, cqe need to zero before use. becauese such buffer will reuse and hardward will change some bit of sqe/cqe. so make sqe/cqe clean is necessary before use.

Reply:
malloc has been replaced to calloc